File: docs/chapters/20-distributed-systems-fundamentals/go/bounded_in_flight.go
// Scenario: Checkout API limits concurrent payment calls during flash sale
// Demonstrates: buffered channel as acquire slot — 503 when pool full
// Trade-off: channel vs semaphore — same backpressure semantics in Go

package main

import (
	"context"
	"errors"
	"time"
)

var errOverloaded = errors.New("overloaded")

type BoundedInFlight struct {
	slots chan struct{}
}

func NewBoundedInFlight(max int) *BoundedInFlight {
	return &BoundedInFlight{slots: make(chan struct{}, max)}
}

func (b *BoundedInFlight) Run(ctx context.Context, fn func(context.Context) error) error {
	select {
	case b.slots <- struct{}{}:
		defer func() { <-b.slots }()
		return fn(ctx)
	case <-time.After(50 * time.Millisecond):
		return errOverloaded
	case <-ctx.Done():
		return ctx.Err()
	}
}

File: docs/chapters/20-distributed-systems-fundamentals/java/BoundedInFlight.java
// Scenario: Checkout API limits concurrent payment calls during flash sale
// Demonstrates: semaphore as backpressure — fail fast with 503 when saturated
// Trade-off: reject at edge vs unbounded queue that OOMs the JVM

import java.util.concurrent.Semaphore;
import java.util.concurrent.TimeUnit;

final class BoundedInFlight {
  private final Semaphore slots;

  BoundedInFlight(int maxConcurrent) {
    this.slots = new Semaphore(maxConcurrent);
  }

  <T> T runWithBackpressure(java.util.concurrent.Callable<T> work) throws Exception {
    if (!slots.tryAcquire(50, TimeUnit.MILLISECONDS)) {
      throw new ServiceUnavailableException("overloaded");
    }
    try {
      return work.call();
    } finally {
      slots.release();
    }
  }

  static final class ServiceUnavailableException extends RuntimeException {
    ServiceUnavailableException(String msg) { super(msg); }
  }
}
